File: Python/740_delete-and-earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        points = defaultdict(int)
        max_num = 0
        for num in nums:
            points[num] += num
            max_num = max(max_num, num)

        max_points = [0] * (max_num + 1)  
        max_points[1] = points[1]

        for i in range(2, len(max_points)):
            max_points[i] = max(max_points[i-1],max_points[i-2]+points[i])

        return max_points[max_num]

File: Java/740_delete-and-earn.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        Map<Integer, Integer> freqMap = new HashMap<>();
        for(int num: nums){
            freqMap.put(num, freqMap.getOrDefault(num, 0)+1);
        }
        List<Integer> list = new ArrayList<>(freqMap.keySet());
        int[] keyList = list.stream().mapToInt(Integer::intValue).toArray();
        Arrays.sort(keyList);
        int oneLeft = 0; int twoLeft = 0;
        for(int i =0; i <keyList.length; i++ ){
            int curVal = keyList[i] * freqMap.get(keyList[i]);
            int temp = oneLeft;
            if(i > 0 && keyList[i] == keyList[i-1]+1){
                oneLeft = Math.max(curVal + twoLeft, oneLeft);
                
            }else{
                oneLeft+= curVal;
            }
            twoLeft = temp;
        }
        return oneLeft;
    }
}